[MSE] video's readyState keeps incorrectly switching between

jyavenard opened the pull request:
#### f78ee0bea7233a1479a018b549d71c827e4e6769
<pre>
[MSE] video&apos;s readyState keeps incorrectly switching between
 HAVE_CURRENT_DATA and HAVE_METADATA
 <a href="https://bugs.webkit.org/show_bug.cgi?id=254399">https://bugs.webkit.org/show_bug.cgi?id=254399</a> rdar://107178961

Reviewed by Youenn Fablet.

Stop using the MediaPlayer&apos;s readyState value to determine if the currentTime is progressing.
Instead use MediaSource&apos;s buffered ranges to determine if we have playable
data in the future.
readyState is asynchronously set and decoupled from currentMediaTime since bug
225396 and may not be up to date at the time we check the currentMediaTime.

Additionally, do not update the cached time when pausing if it&apos;s marked
as not-progressing.

Testing covered in media/media-source/media-source-monitor-source-buffers.html
once webkit.org/b/225367 is done.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::currentTimeMayProgress const):
(WebCore::MediaPlayer::readyState const):
(WebCore::MediaPlayer::readyState): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp: Added.
(WebCore::MediaSourcePrivate::hasFutureTime const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::setTimeFudgeFactor):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::currentMediaTimeMayProgress const):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerPausedOrStalled const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::pause):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):

Canonical link: <a href="https://commits.webkit.org/262112@main">https://commits.webkit.org/262112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b53dcc377e770ebe31ca264768c05116a788d210

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/615 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/764 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/536 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/607 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/681 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/774 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/586 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/752 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/572 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/606 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/548 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/594 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/580 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/587 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/142 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/590 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
```diff
@@ -111,6 +111,7 @@ class MediaPlayerPrivateMediaSourceAVFObjC
     void characteristicsChanged();
 
     MediaTime currentMediaTime() const override;
+    bool currentMediaTimeMayProgress() const override;
```
final?

jyavenard replied:
ok

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -472,6 +473,11 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
 
+bool MediaPlayerPrivateMediaSourceAVFObjC::currentMediaTimeMayProgress() const
+{
+    return MediaSourcePrivate::hasFutureTime(currentMediaTime(), durationMediaTime(), *buffered());
```
How do we know buffered() is not null?
It is also a bit sad to allocate some memory in the heap to just free it right away.
optional<> or Markable would be better there.

jyavenard replied:
I will change the prototyping of this method , it's annoying me too

youennf commented on Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
```diff
@@ -205,6 +206,11 @@ MediaTime MockMediaPlayerMediaSource::currentMediaTime() const
 
+bool MockMediaPlayerMediaSource::currentMediaTimeMayProgress() const
+{
+    return MediaSourcePrivate::hasFutureTime(currentMediaTime(), durationMediaTime(), *buffered());
```
Ditto, can we move this method to the base class?

jyavenard replied:
There's already a default implementation, which is based on the readyState.

youennf commented on Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
```diff
@@ -51,6 +51,7 @@ class MockMediaPlayerMediaSource : public MediaPlayerPrivateInterface, public Ca
 
     void advanceCurrentTime();
     MediaTime currentMediaTime() const override;
+    bool currentMediaTimeMayProgress() const override;
```
final

jyavenard replied:
there's heaps of them marked as override, I thought it was better to keep the style , even if incorrect

youennf commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
```diff
@@ -121,6 +121,7 @@ class MediaPlayerPrivateInterface {
     virtual float currentTime() const { return 0; }
     virtual double currentTimeDouble() const { return currentTime(); }
     virtual MediaTime currentMediaTime() const { return MediaTime::createWithDouble(currentTimeDouble()); }
+    virtual bool currentMediaTimeMayProgress() const { return readyState() >= MediaPlayer::ReadyState::HaveFutureData; }
```
It is strange to use something that is not working as the default.
Should we use virtual pure, or the default non remote implementation and and ASSERT NOT REACHED in the remote implementation?

jyavenard replied:
the buffered range for CRABS playback is an approximation at best. For non-MSE you can never expect a gap in the content to be behaving in a predictive manner as this condition is almost always transitional. Unlike MSE which requires the player.

The only reason the time is readyState is moving back and forth is a consequence of having the MediaSource object constantly updating the readyState, something that wouldn't happen with CRABS.

So while I don't have a definitive solution for this problem, in practice it doesn't matter and it's no worse than using the buffered range

jernoble commented on Source/WebCore/platform/graphics/MediaSourcePrivate.h:
```diff
@@ -69,13 +69,12 @@ class MediaSourcePrivate : public RefCounted<MediaSourcePrivate> {
 
+    WEBCORE_EXPORT static const MediaTime& currentTimeFudgeFactor();
+    WEBCORE_EXPORT static bool hasFutureTime(const MediaTime& currentTime, const MediaTime& duration, const PlatformTimeRanges&);
+
```
I would rather not make the "currentTimeFudgeFactor()" static global. I suspect we will have to pull this back into non-static scope in the relatively near term if the MSE v2 spec settles on a solution for allowing this value to be specified by the page.

jyavenard replied:
ok. Will keep it then.

jyavenard committed:
[MSE] video's readyState keeps incorrectly switching between
 HAVE_CURRENT_DATA and HAVE_METADATA
 https://bugs.webkit.org/show_bug.cgi?id=254399 rdar://107178961

Reviewed by Youenn Fablet.

Stop using the MediaPlayer's readyState value to determine if the currentTime is progressing.
Instead use MediaSource's buffered ranges to determine if we have playable
data in the future.
readyState is asynchronously set and decoupled from currentMediaTime since bug
225396 and may not be up to date at the time we check the currentMediaTime.

Additionally, do not update the cached time when pausing if it's marked
as not-progressing.

Testing covered in media/media-source/media-source-monitor-source-buffers.html
once webkit.org/b/225367 is done.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::currentTimeMayProgress const):
(WebCore::MediaPlayer::readyState const):
(WebCore::MediaPlayer::readyState): Deleted.
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp: Added.
(WebCore::MediaSourcePrivate::hasFutureTime const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::setTimeFudgeFactor):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::currentMediaTimeMayProgress const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::currentMediaTimeMayProgress const):
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerPausedOrStalled const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::pause):
(WebKit::MediaPlayerPrivateRemote::currentTimeChanged):

Canonical link: https://commits.webkit.org/262112@main
